query-service-impl/src/main/java/org/hypertrace/core/query/service/projection/ProjectionTransformation.java

```diff
@@ -315,6 +323,77 @@ private Filter rebuildFilterOmittingDefaults(
     return builder.clearChildFilter().addAllChildFilter(childFilters).build();
   }
 
+  private QueryRequest rebuildRequest(
+      QueryRequest original,
+      List<Expression> selections,
+      List<Expression> aggregations,
+      Filter filter,
+      List<Expression> groupBys,
+      List<OrderByExpression> orderBys) {
+
+    Set<AttributeExpression> attributeExpressionSet =
+        Stream.concat(
+                findComplexAttributeExpressionFromOrderBy(orderBys),
+                findComplexAttributeExpressionFromFilter(filter))
+            .collect(Collectors.toSet());
+
+    return rebuildRequestOmittingDefaults(
+        original,
+        selections,
+        aggregations,
+        createFilterForComplexAttributeExpression(attributeExpressionSet, filter),
+        groupBys,
+        orderBys);
+  }
+
+  private Stream<AttributeExpression> findComplexAttributeExpressionFromOrderBy(
+      List<OrderByExpression> orderByExpressionList) {
+    Stream.Builder<AttributeExpression> builder = Stream.builder();
```

Comment:
The stream is an alternative to an imperative approach. Using a functional paradigm like streams often (not always) makes the code easier to read in my opinion, but combining using streams with an imperative for loop combines the problems of both. Writing this method with streams and a functional style might look like

```java
return orderByExpressionList.stream()
  .filter(QueryRequestUtil::isMapSubpathAttributeExpression) // move this into some util since it's shared logic
  .map(OrderByExpression::getExpression)
  .map(Expression::getAttributeExpression)
  .collect(Collectors.toUnmodifiableSet);
```

Reply:
Passing streams across functions on the other hand is a bit more risky (streams can only be iterated once), so it usually makes sense to keep that as an impl detail rather than exposing it as the return type.

Reply:
resolved